crates/sui-core/src/execution_cache/proxy_cache.rs

```diff
@@ -0,0 +1,422 @@
+use sui_types::messages_checkpoint::CheckpointSequenceNumber;
+use sui_types::object::Object;
+use sui_types::storage::{MarkerValue, ObjectKey, ObjectOrTombstone, PackageObject};
+use sui_types::sui_system_state::SuiSystemState;
+use sui_types::transaction::{VerifiedSignedTransaction, VerifiedTransaction};
+
+use super::{
+    CheckpointCache, ExecutionCacheCommit, ExecutionCacheConfigType, ExecutionCacheMetrics,
+    ExecutionCacheReconfigAPI, ExecutionCacheWrite, ObjectCacheRead, PassthroughCache,
+    StateSyncAPI, TestingAPI, TransactionCacheRead, WritebackCache,
+};
+
+macro_rules! delegate_method {
+    ($self:ident.$method:ident($($args:ident),*)) => {
+        match *$self.mode.lock() {
+            ExecutionCacheConfigType::PassthroughCache => $self.passthrough_cache.$method($($args),*),
+            ExecutionCacheConfigType::WritebackCache => $self.writeback_cache.$method($($args),*),
+        }
+    };
+}
+
+pub struct ProxyCache {
+    // Note: both caches must be constructed at startup, rather than using ArcSwap
+    // (or some similar strategy). This is because we need to proxy iter_live_object_set,
+    // which requires that we borrow from a member of the cache. If we used ArcSwap,
+    // we would be forced to borrow from a local variable after loading from the ArcSwap.
+    //
+    // Cache implementations are entirely passive, so the unused one will have no effect.
+    passthrough_cache: PassthroughCache,
+    writeback_cache: WritebackCache,
+    mode: Mutex<ExecutionCacheConfigType>,
+}
+
+impl ProxyCache {
+    pub fn new(
+        epoch_start_config: &EpochStartConfiguration,
+        store: Arc<AuthorityStore>,
+        metrics: Arc<ExecutionCacheMetrics>,
+    ) -> Self {
+        let cache_type = epoch_start_config.execution_cache_type();
+        tracing::info!("using cache impl {:?}", cache_type);
+        let passthrough_cache = PassthroughCache::new(store.clone(), metrics.clone());
+        let writeback_cache = WritebackCache::new(store.clone(), metrics.clone());
+
+        Self {
+            passthrough_cache,
+            writeback_cache,
+            mode: Mutex::new(cache_type),
+        }
+    }
+
+    async fn reconfigure_cache_impl(&self, epoch_start_config: &EpochStartConfiguration) {
+        let cache_type = epoch_start_config.execution_cache_type();
+        tracing::info!("switching to cache impl {:?}", cache_type);
+        if matches!(cache_type, ExecutionCacheConfigType::PassthroughCache) {
+            // we may switch back to the writeback cache next epoch, at which point its caches will
+            // be stale if not cleared now
+
+            // When we call invalidate_all on Moka caches, it sets the valid after time stamp to the current
+            // time. Upon retrieval, it ignores entries whose insertion time is strictly less than the valid-after
+            // time. In the simulator, time remains constant for the duration of a single task poll, so it is
```

Comment:
Yes, all normal transaction execution is excluded from happening during reconfig by the execution lock. If it wasn't it would be unsafe even without the cache. This is roughly the same as `yield_now()` but I want to make absolutely sure that the clock will have advanced before we call invalidate. (Even though I don't think it would be possible in practice for it to not advance when calling `yield_now()`)

Reply:
Sounds good!